File: app/src/main/java/com/godnit/taskcity/TaskStore.java
package com.godnit.taskcity;

import android.content.Context;
import android.content.SharedPreferences;

import org.json.JSONArray;
import org.json.JSONException;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;

final class TaskStore {
    private static final String PREFS = "task_city_data";
    private static final String KEY_TASKS = "tasks";
    private static final String KEY_NORMAL_HOUSES = "normal_houses";
    private static final String KEY_URGENT_HOUSES = "urgent_houses";

    private final SharedPreferences preferences;
    private final List<TaskItem> tasks = new ArrayList<>();
    private int normalHouses;
    private int urgentHouses;

    TaskStore(Context context) {
        preferences = context.getSharedPreferences(PREFS, Context.MODE_PRIVATE);
        normalHouses = preferences.getInt(KEY_NORMAL_HOUSES, 0);
        urgentHouses = preferences.getInt(KEY_URGENT_HOUSES, 0);
        loadTasks();
    }

    List<TaskItem> tasksOfType(int type) {
        List<TaskItem> filtered = new ArrayList<>();
        for (TaskItem task : tasks) {
            if (task.type == type) filtered.add(task);
        }
        return filtered;
    }

    void add(String title, int type, long dueAt) {
        tasks.add(new TaskItem(System.nanoTime(), title, type, dueAt));
        save();
    }

    TaskItem complete(long id) {
        Iterator<TaskItem> iterator = tasks.iterator();
        while (iterator.hasNext()) {
            TaskItem task = iterator.next();
            if (task.id == id) {
                iterator.remove();
                if (task.type == TaskItem.NORMAL) normalHouses++;
                else urgentHouses++;
                save();
                return task;
            }
        }
        return null;
    }

    int expireUrgentTasks(long now) {
        int expired = 0;
        Iterator<TaskItem> iterator = tasks.iterator();
        while (iterator.hasNext()) {
            TaskItem task = iterator.next();
            if (task.type == TaskItem.URGENT && task.dueAt > 0 && task.dueAt <= now) {
                iterator.remove();
                if (urgentHouses > 0) urgentHouses--;
                expired++;
            }
        }
        if (expired > 0) save();
        return expired;
    }

    int getHouseCount(int type) {
        return type == TaskItem.NORMAL ? normalHouses : urgentHouses;
    }

    private void loadTasks() {
        String json = preferences.getString(KEY_TASKS, "[]");
        try {
            JSONArray array = new JSONArray(json);
            for (int i = 0; i < array.length(); i++) {
                tasks.add(TaskItem.fromJson(array.getJSONObject(i)));
            }
        } catch (JSONException ignored) {
            tasks.clear();
        }
    }

    private void save() {
        JSONArray array = new JSONArray();
        for (TaskItem task : tasks) {
            try {
                array.put(task.toJson());
            } catch (JSONException ignored) {
                // This model contains only primitive values, so serialization is safe.
            }
        }
        preferences.edit()
                .putString(KEY_TASKS, array.toString())
                .putInt(KEY_NORMAL_HOUSES, normalHouses)
                .putInt(KEY_URGENT_HOUSES, urgentHouses)
                .apply();
    }
}

File: app/src/main/java/com/godnit/taskcity/TaskStore.kt
package com.godnit.taskcity

import android.content.Context
import org.json.JSONArray
import org.json.JSONObject
import java.util.UUID

/**
 * Tiny persistent store. Everything lives locally on the phone so timers and
 * the city survive app restarts and periods without Internet access.
 */
class TaskStore(context: Context) {
    data class CityTask(
        val id: String,
        val title: String,
        val createdAt: Long,
        val endAt: Long
    )

    data class ExpireResult(val expiredTasks: Int, val destroyedHouses: Int)

    private val prefs = context.getSharedPreferences("task_city_state", Context.MODE_PRIVATE)
    private val lock = Any()

    fun getTasks(): List<CityTask> = synchronized(lock) {
        readTasks().sortedBy { it.endAt }
    }

    fun addTask(title: String, durationMillis: Long): CityTask = synchronized(lock) {
        val now = System.currentTimeMillis()
        val task = CityTask(
            id = UUID.randomUUID().toString(),
            title = title.trim(),
            createdAt = now,
            endAt = now + durationMillis.coerceAtLeast(60_000L)
        )
        val tasks = readTasks().toMutableList()
        tasks += task
        writeTasks(tasks)
        task
    }

    fun completeTask(id: String): Boolean = synchronized(lock) {
        val tasks = readTasks().toMutableList()
        val removed = tasks.removeAll { it.id == id }
        if (!removed) return@synchronized false
        writeTasks(tasks)
        houseCount = houseCount + 1
        true
    }

    fun expireDueTasks(now: Long = System.currentTimeMillis()): ExpireResult = synchronized(lock) {
        val tasks = readTasks()
        val expired = tasks.filter { it.endAt <= now }
        if (expired.isEmpty()) return@synchronized ExpireResult(0, 0)

        writeTasks(tasks.filter { it.endAt > now })
        val before = houseCount
        houseCount = (before - expired.size).coerceAtLeast(0)
        ExpireResult(expired.size, (before - houseCount).coerceAtLeast(0))
    }

    var houseCount: Int
        get() = prefs.getInt(KEY_HOUSES, DEFAULT_DEMO_HOUSES)
        set(value) {
            prefs.edit().putInt(KEY_HOUSES, value.coerceAtLeast(0)).apply()
        }

    private fun readTasks(): List<CityTask> {
        val raw = prefs.getString(KEY_TASKS, "[]") ?: "[]"
        return try {
            val array = JSONArray(raw)
            val result = ArrayList<CityTask>(array.length())
            for (i in 0 until array.length()) {
                val item = array.getJSONObject(i)
                result += CityTask(
                    id = item.optString("id"),
                    title = item.optString("title", "مهمة"),
                    createdAt = item.optLong("createdAt"),
                    endAt = item.optLong("endAt")
                )
            }
            result.filter { it.id.isNotBlank() && it.endAt > 0L }
        } catch (_: Throwable) {
            emptyList()
        }
    }

    private fun writeTasks(tasks: List<CityTask>) {
        val array = JSONArray()
        tasks.forEach { task ->
            array.put(JSONObject().apply {
                put("id", task.id)
                put("title", task.title)
                put("createdAt", task.createdAt)
                put("endAt", task.endAt)
            })
        }
        prefs.edit().putString(KEY_TASKS, array.toString()).apply()
    }

    companion object {
        private const val KEY_TASKS = "tasks_json"
        private const val KEY_HOUSES = "house_count"
        // A small starter neighborhood so the first prototype immediately shows
        // the visual direction. In the final release this can start at zero.
        private const val DEFAULT_DEMO_HOUSES = 6
    }
}
